jvm/core/src/main/scala/org/scalatest/concurrent/Futures.scala

```diff
@@ -600,9 +605,37 @@ trait Futures extends PatienceConfiguration {
    *          <code>interval</code> parameters that are unused by this method
    * @return the result of invoking the <code>fun</code> parameter
    */
-  final def whenReady[T, U](future: FutureConcept[T])(fun: T => U)(implicit config: PatienceConfig, pos: source.Position): U = {
-    val result = future.futureValueImpl(pos)(config)
+  // SKIP-DOTTY-START
+  final def whenReady[T, U](future: FutureConcept[T])(fun: T => U)(implicit config: PatienceConfig, pos: source.Position): U = 
+    Futures.whenReadyImpl(this)(future, fun, config.timeout, config.interval, pos)
+  // SKIP-DOTTY-END
+  //DOTTY-ONLY final inline def whenReady[T, U](future: FutureConcept[T])(fun: T => U)(implicit config: PatienceConfig, pos: source.Position): U = 
+  //DOTTY-ONLY   ${ Futures.whenReadyMacro('{this}, '{future}, '{fun}, '{config.timeout}, '{config.interval}) }
+
+  /**/
+}
+
+object Futures extends Futures {
+
+  final def whenReadyImpl[T, U](futures: Futures)(future: futures.FutureConcept[T], fun: T => U, timeout: Span, interval: Span, pos: source.Position): U = {
+    val result = future.futureValueImpl(pos)(futures.PatienceConfig(timeout, interval))
     fun(result)
   }
-}
 
+  //DOTTY-ONLY import scala.quoted._
+
+  //DOTTY-ONLY final def workaroundWhenReadyImpl[T, U](futures: Futures)(future: Any, fun: Any, timeout: Span, interval: Span, pos: source.Position): U = 
+  //DOTTY-ONLY   whenReadyImpl(futures)(future.asInstanceOf[futures.FutureConcept[T]], fun.asInstanceOf[T => U], timeout, interval, pos)
+
+  //DOTTY-ONLY // Ideally, we can use future: Expr[futures.FutureConcept[T]] and fun Expr[T => U] here, can't get it to work so we have the above workaroundWhenReadyImpl that takes Any.
+  //DOTTY-ONLY private[concurrent] def whenReadyMacro[T, U](futures: Expr[Futures], future: Expr[Any], fun: Expr[Any], timeout: Expr[Span], interval: Expr[Span])(using quotes: Quotes): Expr[U] = {
```

Comment:
@liufengyun I am trying to use a dependent type here for future argument, which I tried future: Expr[futures.FutureConcept] but it did't work.  Also for the 'fun' argument it is supposed to be Expr[T => U] but it didn't work also, is there a correct way to make dependent type to work?

Cheers.

Reply:
@cheeseng You need `inline def whenReady[T: Type, U: Type]` in the macro definition for `Expr[T => U]` to work.

Path-dependent types like `Expr[futures.FutureConcept]` will not work, as the parameter `futures` has the type `Expr[Futures]` instead of `Futures`. However, you can use `Expr[Futures#FutureConcept]`, which is a little better than `Expr[Any]`. I don't see how to do better than this.